### PR TITLE
empty from for gas estimate [WIP]

### PR DESCRIPTION
**Pull request by thomas-nguy**
Closes: #201

## Description

Allows empty from field when simulating transaction

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/keeper/state_transition.go [hunk 1]
```diff
@@ -121,10 +121,9 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 	// get the latest signer according to the chain rules from the config
 	signer := ethtypes.MakeSigner(ethCfg, big.NewInt(k.ctx.BlockHeight()))
 
-	msg, err := tx.AsMessage(signer)
-	if err != nil {
-		return nil, stacktrace.Propagate(err, "failed to return ethereum transaction as core message")
-	}
+	// Ignore the error because the signature and field from is not needed in case of simulate
+	// For non simulate tx, this should be already validated through the ante handler
+	msg, _ := tx.AsMessage(signer)
```
Comment (yihuang) [inline x/evm/keeper/state_transition.go:126]: Inside `AsMessage` it'll set `from` to the one returned by signer.

**Review by leejw51crypto — APPROVED** [review]
lgtm

File: app/ante/eth.go [hunk 2]
```diff
@@ -187,6 +195,10 @@ func NewEthNonceVerificationDecorator(ak AccountKeeper) EthNonceVerificationDeco
 // AnteHandle validates that the transaction nonces are valid and equivalent to the sender account’s
 // current nonce.
 func (nvd EthNonceVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if simulate {
+		return next(ctx, tx, simulate)
+	}
+
 	// no need to check the nonce on ReCheckTx
 	if ctx.IsReCheckTx() {
```
Comment (fedekunze) [inline app/ante/eth.go:203]:
```suggestion
	// no need to check the nonce on ReCheckTx or simulation
	if simulate || ctx.IsReCheckTx() {
```

File: x/evm/keeper/state_transition.go (hunk 1 above)
Comment (fedekunze) [reply x/evm/keeper/state_transition.go:126]: what if `msg` is `nil`?

Comment (thomas-nguy) [reply x/evm/keeper/state_transition.go:126]: `msg` can't be nil is the tx is not nil because the object is created in `AsMessage`

**Comment by fedekunze** [comment]
is this PR still WIP?

**Comment by thomas-nguy** [comment]
Yes, 

We needs to be able the "gas refund" in case of simulate in `state_transition.go`. I havent found a way to do it yet>
Maybe #236 might help

**Comment by yihuang** [comment]
I think https://github.com/tharsis/ethermint/pull/236 fixes this empty from field issue, according to my test.

**Comment by thomas-nguy** [comment]
Ok closing the PR as it will be fixed in #236

**Pull request closed without merging**